src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
+  }
+
+  // Propagate matrix, but only if we are the root.
+  if (parent == NULL)
+  {
+    std::queue<CosineTree*> queue;
+    if (left)
+      queue.push(left);
+    if (right)
+      queue.push(right);
+    while (!queue.empty())
+    {
+      CosineTree* node = queue.front();
+      queue.pop();
+
+      node->dataset = dataset;
+      if (node->left)
+        queue.push(node->left);
+      if (node->right)
+        queue.push(node->right);
+    }
+  }
+}
+
+//! Copy assignment operator: copy the given other tree.
+CosineTree& CosineTree::operator=(const CosineTree& other)
+{
+  // Return if it's the same tree.
+  if (this == &other)
+    return *this;
+
+  // Freeing memory that will not be used anymore.
+  delete left;
+  delete right;
+
+  dataset = (other.parent == NULL) ? other.parent->GetDataset() : NULL;
+  delta = other.delta;
+  parent = other.Parent();
+  left = other.Left();
+  right = other.Right();
+  indices = other.indices;
+  l2NormsSquared = other.l2NormsSquared;
+  centroid = other.centroid;
+  basisVector = other.basisVector;
+  splitPointIndex = other.SplitPointIndex();
+  numColumns = other.NumColumns();
+  l2Error = other.L2Error();
+  frobNormSquared = other.FrobNormSquared();
+
+  // Create left and right children (if any).
+  if (other.Left())
+  {
+    left = new CosineTree(*other.Left());
+    left->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  if (other.Right())
+  {
+    right = new CosineTree(*other.Right());
+    right->Parent() = this; // Set parent to this, not other tree.
+  }
+
+  // Propagate matrix, but only if we are the root.
+  if (parent == NULL)
+  {
+    std::queue<CosineTree*> queue;
+    if (left)
+      queue.push(left);
+    if (right)
+      queue.push(right);
+    while (!queue.empty())
+    {
+      CosineTree* node = queue.front();
+      queue.pop();
+
+      node->dataset = dataset;
+      if (node->left)
+        queue.push(node->left);
+      if (node->right)
+        queue.push(node->right);
+    }
+  }
+
+  return *this;
+}
+
+//! Move the given tree.
+CosineTree::CosineTree(CosineTree&& other) :
+    dataset(other.dataset),
+    delta(std::move(other.delta)),
+    parent(other.parent),
+    left(other.left),
+    right(other.right),
+    indices(std::move(other.indices)),
+    l2NormsSquared(std::move(other.l2NormsSquared)),
+    centroid(std::move(other.centroid)),
+    basisVector(std::move(other.basisVector)),
+    splitPointIndex(other.splitPointIndex),
+    numColumns(other.numColumns),
+    l2Error(other.l2Error),
+    frobNormSquared(other.frobNormSquared)
+{
+  // Now we are a clone of the other tree.  But we must also clear the other
+  // tree's contents, so it doesn't delete anything when it is destructed.
+  other.dataset = NULL;
+  other.parent = NULL;
+  other.left = NULL;
+  other.right = NULL;
+  other.splitPointIndex = 0;
+  other.numColumns = dataset->n_cols;
+  other.l2Error = -1;
+  other.frobNormSquared = arma::accu(l2NormsSquared);
```

Comment:
It doesn't make sense to me why you are setting parts of `other` to have values relevant to this tree.